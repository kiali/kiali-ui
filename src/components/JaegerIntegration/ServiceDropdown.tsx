import * as React from 'react';
import { connect } from 'react-redux';
import { KialiAppState } from '../../store/Store';
import { FormSelect, FormSelectOption, FormSelectOptionGroup } from '@patternfly/react-core';
import * as Api from '../../services/Api';
import { PromisesRegistry } from '../../utils/CancelablePromises';
import { ServiceOverview } from '../../types/ServiceList';
import { style } from 'typestyle';

interface ServiceDropdownProps {
  disabled?: boolean;
  activeNamespaces: string[];
  service?: string;
  setService: (service: string) => void;
}

interface Service {
  value: string;
  label: string;
  disabled: boolean;
}

interface ServiceGroup {
  groupLabel: string;
  disabled: boolean;
  options: Service[];
}
interface ServiceDropdownState {
  servicesGroups: ServiceGroup[];
}

const serviceDropdown = style({ marginLeft: '-100px' });

export class ServiceDropdown extends React.PureComponent<ServiceDropdownProps, ServiceDropdownState> {
  constructor(props: ServiceDropdownProps) {
    super(props);
    this.state = { servicesGroups: [] };
    if (this.props.activeNamespaces.length > 0) {
      this.refreshServices(this.props.activeNamespaces);
    }
  }

  componentDidUpdate(prevProps: ServiceDropdownProps) {
    if (prevProps.activeNamespaces.sort().join(',') !== this.props.activeNamespaces.sort().join(',')) {
      this.refreshServices(this.props.activeNamespaces);
    }
  }

  refreshServices = (namespaces: string[]) => {
    if (namespaces.length === 0) {
      this.setState({ servicesGroups: [] });
    } else {
      const servicesPromises = namespaces.map(ns => Api.getServices(ns));
      const promises = new PromisesRegistry();
      promises
        .registerAll('services', servicesPromises)
        .then(responses => {
          const serviceList: ServiceGroup[] = [];
          responses.forEach(response => {
            const ns = response.data.namespace.name;
            const serviceGroup: ServiceGroup = { groupLabel: ns, disabled: false, options: [] };

            response.data.services.forEach((service: ServiceOverview) => {
              serviceGroup.options.push({ value: `${service.name}.${ns}`, label: service.name, disabled: false });
            });
            serviceList.push(serviceGroup);
          });
          this.setState({ servicesGroups: serviceList });
        })
        .catch(() => console.log('Error'));
    }
  };

  handleFocus = () => this.refreshServices(this.props.activeNamespaces);

  labelServiceDropdown = (items: number) => {
    if (this.props.activeNamespaces.length > 0) {
      if (items === 0) {
        return 'Select another namespace with services';
      }
      return 'Select a service';
    }
    return 'Select a namespace';
  };

  render() {
    const { disabled } = this.props;
    const { servicesGroups } = this.state;

    return (
      <FormSelect
        value={this.props.service}
        isDisabled={disabled || this.props.activeNamespaces.length === 0 || Object.keys(servicesGroups).length === 0}
        onFocus={() => this.handleFocus}
        onChange={this.props.setService}
        aria-label="FormSelect Input"
        className={serviceDropdown}
      >
        <FormSelectOption
          isDisabled={false}
          key={'help_test'}
          value={''}
          label={this.labelServiceDropdown(Object.keys(servicesGroups).length)}
        />
        {servicesGroups.map((group, index) => (
          <FormSelectOptionGroup isDisabled={group.disabled} key={index} label={group.groupLabel}>
            {group.options.map((option, i) => (
              <FormSelectOption isDisabled={option.disabled} key={i} value={option.value} label={option.label} />
            ))}
          </FormSelectOptionGroup>
        ))}
      </FormSelect>
    );
  }
}

const mapStateToProps = (state: KialiAppState) => {
  return {
    activeNamespaces: state.namespaces.activeNamespaces.map(ns => ns.name)
  };
};

const ServiceDropdownContainer = connect(mapStateToProps)(ServiceDropdown);

export default ServiceDropdownContainer;
