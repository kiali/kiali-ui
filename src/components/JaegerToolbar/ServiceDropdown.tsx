import * as React from 'react';
import { connect } from 'react-redux';
import { KialiAppState } from '../../store/Store';
import ToolbarDropdown from '../../components/ToolbarDropdown/ToolbarDropdown';
import { JaegerActions } from '../../actions/JaegerActions';
import { JaegerThunkActions } from '../../actions/JaegerThunkActions';
import { ThunkDispatch } from 'redux-thunk';
import { KialiAppAction } from '../../actions/KialiAppAction';

interface ServiceDropdownProps {
  disabled: boolean;
  activeNamespace: string;
  service: string;
  items: string[];
  refresh: (ns: string) => void;
  setService: (service: string) => void;
}

export class ServiceDropdown extends React.PureComponent<ServiceDropdownProps, {}> {
  constructor(props: ServiceDropdownProps) {
    super(props);
  }

  componentDidMount() {
    if (this.props.activeNamespace) {
      this.props.refresh(this.props.activeNamespace);
    }
  }

  componentDidUpdate(prevProps: ServiceDropdownProps) {
    if (this.props.activeNamespace !== prevProps.activeNamespace && this.props.activeNamespace) {
      this.props.refresh(this.props.activeNamespace);
    }
  }

  handleToggle = (isOpen: boolean) => isOpen && this.props.refresh(this.props.activeNamespace);

  labelServiceDropdown = (items: number) => {
    if (this.props.activeNamespace && this.props.activeNamespace !== 'all') {
      if (items === 0) {
        return 'Choose another namespace with services';
      }
      return 'Choose a service';
    }
    return 'Choose a namespace';
  };

  render() {
    const { disabled } = this.props;

    const items: { [key: string]: string } = this.props.items.reduce((list, item) => {
      list[item] = item;
      return list;
    }, {});

    return (
      <span style={{ marginLeft: '10px' }}>
        <ToolbarDropdown
          id="namespace-selector"
          disabled={disabled || Object.keys(items).length === 0}
          options={items}
          value={''}
          label={this.props.service || this.labelServiceDropdown(Object.keys(items).length)}
          useName={true}
          handleSelect={this.props.setService}
          onToggle={this.handleToggle}
        />
      </span>
    );
  }
}

const mapStateToProps = (state: KialiAppState) => {
  return {
    items: state.jaegerState.toolbar.services,
    disabled: state.jaegerState.toolbar.isFetchingService,
    activeNamespace: state.jaegerState.search.namespaceSelected,
    service: state.jaegerState.search.serviceSelected
  };
};

const mapDispatchToProps = (dispatch: ThunkDispatch<KialiAppState, void, KialiAppAction>) => {
  return {
    refresh: (ns: string) => {
      dispatch(JaegerThunkActions.asyncFetchServices(ns));
    },
    setService: (service: string) => {
      dispatch(JaegerActions.setService(service));
    }
  };
};

const ServiceDropdownContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(ServiceDropdown);

export default ServiceDropdownContainer;
