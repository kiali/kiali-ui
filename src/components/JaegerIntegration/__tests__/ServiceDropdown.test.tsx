import * as React from 'react';
import { shallow } from 'enzyme';
import { ServiceDropdown } from '../ServiceDropdown';
import { shallowToJson } from 'enzyme-to-json';

describe('NamespaceDropdown', () => {
  let wrapper, setService;
  beforeEach(() => {
    setService = jest.fn();
    wrapper = shallow(<ServiceDropdown setService={setService} service={''} activeNamespaces={[]} disabled={false} />);
  });

  it('renders ServiceDropdown correctly without custom', () => {
    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
