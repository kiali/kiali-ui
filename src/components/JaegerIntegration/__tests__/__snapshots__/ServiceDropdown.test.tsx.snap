// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NamespaceDropdown renders ServiceDropdown correctly without custom 1`] = `
<FormSelect
  aria-label="FormSelect Input"
  className="f6vm0mb"
  isDisabled={true}
  isValid={true}
  onBlur={[Function]}
  onChange={[MockFunction]}
  onFocus={[Function]}
  value=""
>
  <FormSelectOption
    isDisabled={false}
    key="help_test"
    label="Select a namespace"
    value=""
  />
</FormSelect>
`;
